### DISCUSSION TITLE
Having trouble running Application within another running loop...

### DISCUSSION THREAD

**Discussion by tumma72** [python-telegram-bot/python-telegram-bot#3310] [Q&A] [answered]
Hi there,
 first compliments for the amazing work everyone did with PTB, it is really well thought through, and very well coded! I have recently started my attempt to integrate the PTB into another application which is already managing the event loop. My intention was to synchronize sending messages on events (which is quite easy using Bot.send_message()) with updating the message queue at the end of every loop. Now without being able to start with Application.run_poll() unless considering going multiple Threads or processes, I was wondering how can I manually get the Application.update_queue to fill and then running the Application.start() to process the queue and get the benefit of the Handlers, without having to write custom code to process the updates.

So far I have managed to use Bot.get_updates() then update the Application.update_queue but when I try to run Application.start() I get a RuntimeError:
`RuntimeError: The connection pool was closed while 1 HTTP requests/responses were still in-flight.`

In order to initialize the Application outside of the run_poll() context, I have manually used:
```
try:
  self.application._check_initialized()

except RuntimeError as e:
  print(f"Initializing... {e}")
  await self.application.initialize()
```
Which seems to execute.

I hope this makes sense, I'll be happy for any advice you can provide.

Best
ANdreaT

**Comment by Bibo-Joshi** [comment, marked as answer]
Hi, glad that you like the library :)

`Application.run_polling` is purely optional. It's bascially a thin wrapper that runs

* `Application.initalize`
* `Application.start`
* `Application.updater.start_polling`
* `Application.updater.stop`
* `Application.stop`
* `Application.shutdown`

and blocks the event loop until the user sends a stop signal. The last part is what makes `run_polling` unsuitable when combining PTB with other asyncio frameworks. In that case, you can just manually call the methods. See also [this example](https://docs.python-telegram-bot.org/en/v20.0a4/examples.html#examples-customwebhookbot), which showcases this in the context of setting up a custom webhook.

**Reply by tumma72** [reply]
Thanks for the quick reply, and for the concise answer, I will try with this approach, even though I was thinking I could initialize the application once, then run the loop, and finally close... something like:

- Application.Initialize()
- Application.start()
- << Main Loop starts >>
- Application.updater.start_polling()
- Application.updater.stop()
- << End Main Loop >>
- Application.stop()
- Application.shutdown()

Meaning I would expect to initialize and start, then within the loop update the queue, then when done stop application and shutdown. I will try out some options and revert.

Thanks so much for pointing me in the right direction :-)
ANdreaT

**Reply by Bibo-Joshi** [reply]
depends a bit on what you mean by "loop" here. All these methods are coroutine methods that schedule/await tasks/open & close asyncio-related resoucres etc, so they will all need to be run in the same *event loop* (at least I guess so …). If by "main loop" you just mean "some logic that keeps the script alive as long as I need it", then your logic should be fine …

**Reply by al-matty** [reply]
This discussion has solved an issue for me. This might be helpful for anyone else who is trying to run the telegram bot alongside another async application (in my case a discord bot). This ia what I came up with by combining @Bibo-Joshi's reply and the mentioned implementation in [customwebhookbot.py](https://docs.python-telegram-bot.org/en/v20.0a6/examples.customwebhookbot.html).

```code: python
async def run():

    # build application
    application = Application.builder().token("TOKEN").persistence(persistence).build()

    # add handlers
    conv_handler = ConversationHandler("CONVERSATION HANDLERS GO HERE")
    cmd_handler = CommandHandler("COMMAND HANDLERS GO HERE")
    application.add_handler(conv_handler)
    application.add_handler(cmd_handler)

    # Run application and other_application() within the same event loop
    async with application:
        await application.initialize() # inits bot, update, persistence
        await application.start()
        await application.updater.start_polling()
        await other_application()
```
See repo [here](https://github.com/jediswaplabs/discord-alert-bot).

**Reply by jnxyatmjx** [reply]
I have a problem.
In my situation, i have 2 threads in my program.
get update messages from Telegram  use **application.run_polling()** in main thread. 
and down files in another thread, after down complete ,i use **await app.update_queue.put(VideosUpdate(filename=filename_))** to notify the VideosUpdate .
My question is I get the **VideosUpdate** update maybe 6 secondes later after call  **await app.update_queue.put** . 
why so long time , is there any wrong i got ?

**Reply by jnxyatmjx** [reply]
moreover,
```python
async def downloader(queuq: queue.Queue,app: Application) -> None:
    while True:
        url_ = queuq.get()
        filename_ = dlp_down(url_)
        await app.update_queue.put(VideosUpdate(user_id=9900, filename=filename_))

def downloader_callback(queuq: queue.Queue,app: Application) -> None:
    loop_ = asyncio.new_event_loop()
    asyncio.set_event_loop(loop_)
    loop_.run_until_complete(downloader(queuq ,app))
    loop_.close() 
```
**downloader_callback** run in the downloader thread.

**Reply by Bibo-Joshi** [reply]
I highly recommend not to max asyncio with threading

**Reply by jnxyatmjx** [reply]
OK,
but how combine long time tasks into PTB and not block PTB event loop?

**Reply by jnxyatmjx** [reply]
Thank you very much if there are examples for me

**Reply by Bibo-Joshi** [reply]
Please have a look at https://github.com/python-telegram-bot/python-telegram-bot/wiki/Frequently-requested-design-patterns#running-ptb-alongside-other-asyncio-frameworks and kindly actually read the messages in the thread above. If you have follow-up questions, please open a new thread.
